fix: handle internal notifications during session cleanup

donghao1393 opened the pull request:
# fix: handle internal notifications during session cleanup

## Motivation and Context
Addresses an issue where internal notifications (e.g. 'cancelled') during session cleanup would trigger validation errors and crash the server.

## How Has This Been Tested?
yes

## Breaking Changes
no

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context

### Changes
- Add session state tracking via `_closed` flag
- Skip validation for internal notifications during cleanup phase
- Improve progress context cleanup with proper async context manager
- Add `final_progress` notification to ensure completion state is reported

### Testing
Successfully tested with multiple concurrent requests with progress tracking:
- Cleanup proceeds without validation errors
- Progress notifications complete properly
- No server crashes during shutdown

donghao1393 commented:
The issue also occurred with https://github.com/modelcontextprotocol/servers/tree/main/src/fetch. It was also resolved in this commit.
```
Installed 31 packages in 36ms
(node:61856) ExperimentalWarning: CommonJS module /opt/homebrew/lib/node_modules/npm/node_modules/debug/src/node.js is loading ES Module /opt/homebrew/lib/node_modules/npm/node_modules/supports-color/index.js using require().
Support for loading ES Module in require() is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Traceback (most recent call last):
  File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/anyio/streams/memory.py", line 242, in send
    self.send_nowait(item)
  File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/anyio/streams/memory.py", line 225, in send_nowait
    raise WouldBlock
anyio.WouldBlock

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp/server/__init__.py", line 467, in run
    await message.respond(
  File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp/shared/session.py", line 58, in respond
    await self._session._send_response(
  File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp/shared/session.py", line 196, in _send_response
    await self._write_stream.send(JSONRPCMessage(jsonrpc_error))
  File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/anyio/streams/memory.py", line 248, in send
    await send_event.wait()
  File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 1747, in wait
    await self._event.wait()
  File "/opt/homebrew/Cellar/python@3.12/3.12.7/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/locks.py", line 212, in wait
    await fut
asyncio.exceptions.CancelledError: Cancelled by cancel scope 103cd0650

During handling of the above exception, another exception occurred:

  + Exception Group Traceback (most recent call last):
  |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp/server/stdio.py", line 83, in stdio_server
  |     yield read_stream, write_stream
  |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp_server_fetch/server.py", line 272, in serve
  |     await server.run(read_stream, write_stream, options, raise_exceptions=True)
  |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp/server/__init__.py", line 423, in run
  |     async with ServerSession(
  |                ^^^^^^^^^^^^^^
  |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp/shared/session.py", line 122, in __aexit__
  |     return await self._task_group.__aexit__(exc_type, exc_val, exc_tb)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 763, in __aexit__
  |     raise BaseExceptionGroup(
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp/shared/session.py", line 235, in _receive_loop
    |     notification = self._receive_notification_type.model_validate(
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/pydantic/main.py", line 627, in model_validate
    |     return cls.__pydantic_validator__.validate_python(
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    | pydantic_core._pydantic_core.ValidationError: 5 validation errors for ClientNotification
ProgressNotification.method
  Input should be 'notifications/progress' [type=literal_error, input_value='cancelled', input_type=str]
    For further information visit https://errors.pydantic.dev/2.10/v/literal_error
ProgressNotification.params.progressToken
  Field required [type=missing, input_value={'requestId': 127, 'reaso... -2: Request timed out'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.10/v/missing
ProgressNotification.params.progress
  Field required [type=missing, input_value={'requestId': 127, 'reaso... -2: Request timed out'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.10/v/missing
InitializedNotification.method
  Input should be 'notifications/initialized' [type=literal_error, input_value='cancelled', input_type=str]
    For further information visit https://errors.pydantic.dev/2.10/v/literal_error
RootsListChangedNotification.method
  Input should be 'notifications/roots/list_changed' [type=literal_error, input_value='cancelled', input_type=str]
    For further information visit https://errors.pydantic.dev/2.10/v/literal_error
    +------------------------------------

During handling of the above exception, another exception occurred:

  + Exception Group Traceback (most recent call last):
  |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/bin/mcp-server-fetch", line 10, in <module>
  |     sys.exit(main())
  |              ^^^^^^
  |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp_server_fetch/__init__.py", line 20, in main
  |     asyncio.run(serve(args.user_agent, args.ignore_robots_txt))
  |   File "/opt/homebrew/Cellar/python@3.12/3.12.7/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/runners.py", line 194, in run
  |     return runner.run(main)
  |            ^^^^^^^^^^^^^^^^
  |   File "/opt/homebrew/Cellar/python@3.12/3.12.7/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/runners.py", line 118, in run
  |     return self._loop.run_until_complete(task)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/opt/homebrew/Cellar/python@3.12/3.12.7/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
  |     return future.result()
  |            ^^^^^^^^^^^^^^^
  |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp_server_fetch/server.py", line 271, in serve
  |     async with stdio_server() as (read_stream, write_stream):
  |                ^^^^^^^^^^^^^^
  |   File "/opt/homebrew/Cellar/python@3.12/3.12.7/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 231, in __aexit__
  |     await self.gen.athrow(value)
  |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp/server/stdio.py", line 80, in stdio_server
  |     async with anyio.create_task_group() as tg:
  |                ^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 763, in __aexit__
  |     raise BaseExceptionGroup(
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Exception Group Traceback (most recent call last):
    |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp/server/stdio.py", line 83, in stdio_server
    |     yield read_stream, write_stream
    |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp_server_fetch/server.py", line 272, in serve
    |     await server.run(read_stream, write_stream, options, raise_exceptions=True)
    |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp/server/__init__.py", line 423, in run
    |     async with ServerSession(
    |                ^^^^^^^^^^^^^^
    |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp/shared/session.py", line 122, in __aexit__
    |     return await self._task_group.__aexit__(exc_type, exc_val, exc_tb)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 763, in __aexit__
    |     raise BaseExceptionGroup(
    | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
    +-+---------------- 1 ----------------
      | Traceback (most recent call last):
      |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/mcp/shared/session.py", line 235, in _receive_loop
      |     notification = self._receive_notification_type.model_validate(
      |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      |   File "/Users/donghao/.cache/uv/archive-v0/CiVrAnivdRZCJwWmSC5l6/lib/python3.12/site-packages/pydantic/main.py", line 627, in model_validate
      |     return cls.__pydantic_validator__.validate_python(
      |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      | pydantic_core._pydantic_core.ValidationError: 5 validation errors for ClientNotification
ProgressNotification.method
  Input should be 'notifications/progress' [type=literal_error, input_value='cancelled', input_type=str]
    For further information visit https://errors.pydantic.dev/2.10/v/literal_error
ProgressNotification.params.progressToken
  Field required [type=missing, input_value={'requestId': 127, 'reaso... -2: Request timed out'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.10/v/missing
ProgressNotification.params.progress
  Field required [type=missing, input_value={'requestId': 127, 'reaso... -2: Request timed out'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.10/v/missing
InitializedNotification.method
  Input should be 'notifications/initialized' [type=literal_error, input_value='cancelled', input_type=str]
    For further information visit https://errors.pydantic.dev/2.10/v/literal_error
RootsListChangedNotification.method
  Input should be 'notifications/roots/list_changed' [type=literal_error, input_value='cancelled', input_type=str]
    For further information visit https://errors.pydantic.dev/2.10/v/literal_error
      +------------------------------------
```

dsp-ant reviewed (changes requested):
Thank you for working on this. I think it's the right direction but needs a bit more refinement. If you don't mind taking a look at my requested changes. I think we are close to getting this in.

donghao1393 commented on src/mcp/shared/session.py:
```diff
@@ -204,6 +215,19 @@ async def _send_response(
+                not self._closed
+            )
+        except:
+            return False
```
https://github.com/modelcontextprotocol/python-sdk/pull/85/commits/6c9ca12d32506ddf0e8ead4d216407f389fe432e

donghao1393 commented on src/mcp/shared/session.py:
```diff
@@ -232,14 +256,20 @@ async def _receive_loop(self) -> None:
+                        except Exception:
+                            # Ignore validation errors for notifications during cleanup
+                            if not self._closed:
+                                raise
```
https://github.com/modelcontextprotocol/python-sdk/pull/85/commits/cffdd06464aaea90583847290376b4cc505cd6ca

donghao1393 commented on src/mcp/shared/session.py:
```diff
@@ -204,6 +215,19 @@ async def _send_response(
+        """
+        try:
+            return (
+                getattr(message_root, "method", None) != "cancelled" and
```
https://github.com/modelcontextprotocol/python-sdk/pull/85/commits/e592afd7dba4066cfc115caed754e6b91927ed92

donghao1393 commented:
Hi @dsp-ant , thank you for your advice. I have made commit. Would you help to check again?

donghao1393 commented:
Technical Details
- The issue occurred when internal notifications were received during cleanup
- Added _closed flag to track session state accurately
- Modified notification handling to be cleanup-aware
- Ensures graceful shutdown even with pending notifications

Impact
- Prevents validation errors during shutdown
- Maintains session stability during cleanup
- No interference with normal operation
- Compatible with existing notification patterns

This approach keeps the fix focused on the core issue while maintaining 
backward compatibility.

txbm commented:
@dsp-ant @donghao1393 - if I may suggest a more straight-forward fix for this inline with codebase patterns? - https://github.com/txbm/python-sdk/commit/19185626d57fdd670219784d10859c7d0a302c2e

donghao1393 commented:
> @dsp-ant @donghao1393 - if I may suggest a more straight-forward fix for this inline with codebase patterns? - [txbm@1918562](https://github.com/txbm/python-sdk/commit/19185626d57fdd670219784d10859c7d0a302c2e)

I tested in my local and it's working. I think this is a better solution. What do you think @dsp-ant ?

orliesaurus commented:
Please can you merge this fix in, it's breaking my server on claude desktop app - thank you!

jerome3o-anthropic commented on src/mcp/shared/session.py:
```diff
@@ -204,6 +215,12 @@ async def _send_response(
             )
             await self._write_stream.send(JSONRPCMessage(jsonrpc_response))
 
+    def _is_cancellation_notification(self, message_root: JSONRPCNotification) -> bool:
```
This should no longer be needed now that there are the correct types defined

donghao1393 replied:
Thanks for the suggestions. I share your instinct about simplifying the code based on types, but I have some concerns:

1. The cancellation notification we receive is actually using "cancelled" as the method name, not "notifications/cancelled". This is why we need the explicit check in `_is_cancellation_notification`.

2. Even with the type definitions, the runtime behavior might still need this handling because:
   - The notification comes from the transport layer before type validation
   - We need to handle both formats for backwards compatibility
   - The validation would fail before we can process the notification properly

Would it help if I add some comments explaining this in the code? Or would you prefer to first verify this with the test case I mentioned above?

donghao1393 replied:
removed as no needed

jerome3o-anthropic commented on src/mcp/shared/session.py:
```diff
@@ -232,14 +249,27 @@ async def _receive_loop(self) -> None:
-                    await self._received_notification(notification)
-                    await self._incoming_message_stream_writer.send(notification)
+                    # check if it is a cancellation notification
+                    if self._is_cancellation_notification(message.root):
```
Can this be significantly simplified now that we have the types defined?

donghao1393 replied:
updated with code from https://github.com/modelcontextprotocol/python-sdk/pull/167

jerome3o-anthropic reviewed (changes requested):
Hey @donghao1393 - It seems to me that this can be simplified a lot now that there is the canceled notification types.

I'm going to be doing a loop on notifications in this sdk soon, and for my understanding (and to write some tests) I'd love it if you could also provide an example that reproduces this issue? would it just be a server with a long running task and the client sending of a cancel notification?

donghao1393 commented:
> Hey @donghao1393 - It seems to me that this can be simplified a lot now that there is the canceled notification types.
> 
> I'm going to be doing a loop on notifications in this sdk soon, and for my understanding (and to write some tests) I'd love it if you could also provide an example that reproduces this issue? would it just be a server with a long running task and the client sending of a cancel notification?

Yes, you're right. A reproduction case would be a server with a long-running task (like image generation) that typically takes more than a few seconds, and the client sending a cancel notification during the process. I can provide a minimal example:

1. add this of [my mcp service](https://github.com/donghao1393/mcp-openai/blob/main/README.en.md#setup)
2. A client that initiates the task and sends a cancel notification, like:
    - 1st prompt "create an image using openai mcp service to depict a dog standing on the moon watching the earth"
    - 2nd prompt "create another image that the dog was running"
    - if the cancel notification made the service down in 1st prompt, the 2nd prompt would be failed to send request
4. inspect the log file `mcp-server-openai.log` so that we can observe the cancellation behavior

This would help validate both the current behavior and any proposed changes.

donghao1393 commented:
Hi @jerome3o-anthropic , I have updated accordingly and passed my test. Would you help to take a review for the next?

orliesaurus commented:
I am curious if you could merge this? @dsp-ant 
Maybe it will fix my issues with `The read operation timed out`

donghao1393 commented:
> It seem the issue is fixed, either Claude Desktop update or using the latest main branch. As already I see cancelled got added as valid type, so should no more throw panic error and crash.

I have tested without this PR, but it failed in my case.

donghao1393 commented:
> I have used the main branch. It's not in 1.2.1. For me this seem solving it.
> 
> https://github.com/modelcontextprotocol/python-sdk/blob/f10665db4c2f676da1131617ad67715952258712/src/mcp/types.py#L995
> 
> 
> I testing 1 min task withtout issue when I moved using the new main branch. Previously would crash.
> I had even similar patch as your and removed it.
> Can you test the main branch not 1.2.1.

Thank you for your advice. I tried with
```
# disable old mcp sdk
uv add git+https://github.com/modelcontextprotocol/python-sdk --branch main
# restart claude desktop
```
And then it works.

orliesaurus commented:
> I have used the main branch. It's not in 1.2.1. For me this seem solving it.
> 
> https://github.com/modelcontextprotocol/python-sdk/blob/f10665db4c2f676da1131617ad67715952258712/src/mcp/types.py#L995
> 
> 
> I testing 1 min task withtout issue when I moved using the new main branch. Previously would crash.
> I had even similar patch as your and removed it.
> Can you test the main branch not 1.2.1.

if it's not in 1.2.1 is it going to be out in 1.2.2 or what's the logic you think?
